### ISSUE TITLE
[Error] Peer’s certificate has an invalid signature.

### ISSUE THREAD

**Issue by itsmohamadh** [therealaleph/MasterHttpRelayVPN-RUST#659]
> _**I know my Google or Cloudflare accounts are not flagged** (someone else tested those for me they work)_ 

> _Also checked the certificate in keychain access. It sounds to be installed correctly_ 

I've been getting this in Safari and Chrome:

```
502

Relay error: io: invalid peer certificate: UnknownIssuer`

```

and this in Firefox:
```
https://www.w3schools.com/

Peer’s certificate has an invalid signature.

HTTP Strict Transport Security: false
HTTP Public Key Pinning: false

Certificate chain:

-----BEGIN CERTIFICATE-----
MIIBnzCCAUSgAwIBAgIUTndvcbGARMTBtm2MWan7i/yn10EwCgYIKoZIzj0EAwIw
OjEbMBkGA1UEAwwSTWFzdGVySHR0cFJlbGF5VlBOMRswGQYDVQQKDBJNYXN0ZXJI
dHRwUmVsYXlWUE4wHhcNMjYwNTAzMDU1MTQ4WhcNMjcwNTAzMDU1NjQ4WjAcMRow
*************************************
qmt3+Z3GLcW8WTkrsGCKi7jPIDqjRjBEMBwGA1UdEQQVMBOCEXd3dy53M3NjaG9v
bHMuY29tMA8GA1UdDwEB/wQFAwMHoAAwEwYDVR0lBAwwCgYIKwYBBQUHAwEwCgYI
KoZIzj0EAwIDSQAwRgIhAIoqN7ySSaEBgSnrsE2CHcE5rP4nIrgYzl+iWmOTwize
AiEA5S7lrDMv8Getmvctpf1lrHfQcqGA6nW+nciF80M14Ug=
-----END CERTIFICATE-----
-----BEGIN CERTIFICATE-----
MIIBvDCCAWKgAwIBAgIUDrdG1cQ/7Z1Acml0841JlXs0IX8wCgYIKoZIzj0EAwIw
OjEbMBkGA1UEAwwSTWFzdGVySHR0cFJlbGF5VlBOMRswGQYDVQQKDBJNYXN0ZXJI
dHRwUmVsYXlWUE4wHhcNMjYwNTAzMDU0NzA5WhcNMzYwNDMwMDU1MjA5WjA6MRsw
GQYDVQQDDBJNYXN0ZXJIdHRwUmVsYXlWUE4xGzAZBgNVBAoMEk1hc3Rlckh0dHBS
*************************************
PcajRjBEMA8GA1UdDwEB/wQFAwMHhgAwHQYDVR0OBBYEFFIMSq9nzHQ98IoxA4J7
sY0/PFGJMBIGA1UdEwEB/wQIMAYBAf8CAQAwCgYIKoZIzj0EAwIDSAAwRQIgc3b1
C/v4zmwWOBwJP9f9KfTL4ZSZfonkDMOybWajWHgCIQCH52fsL6P8ZCE8Wu5/xv9W
Plb6W9yWTqpQHIpSmv3eBA==
-----END CERTIFICATE-----

```

Does anybody know where exactly this issue stems from?

**Comment by therealaleph** [comment]
@mhaqnegahdar — `invalid peer certificate` despite "installed correctly" in keychain access usually means one of two things:

**1. Browser cert store ≠ macOS keychain.** Safari uses macOS keychain (so works after `--install-cert`). **Chrome on macOS also uses keychain**. **Firefox uses its OWN cert DB (NSS)** — install runs `certutil` to push the CA into Firefox profiles, but it's best-effort and can silently fail (e.g. Firefox open during install, locked NSS DB, missing certutil). Verify via `Firefox → Settings → Privacy & Security → Certificates → View Certificates → Authorities` — search "MasterHttpRelayVPN". If absent, run:
```bash
brew install nss   # if certutil missing
mhrv-rs --install-cert
```
or import `ca/ca.crt` manually via the Firefox dialog (with Firefox closed first).

**2. Wrong CA on disk.** If you removed `ca/` and started mhrv-rs again, a fresh CA was generated, but the OS / browser still trusts the old (now-deleted) one. Symptom: cert chain validates against a CA that's gone. Fix:
```bash
mhrv-rs --remove-cert    # cleans system trust + on-disk ca/
mhrv-rs --install-cert   # generates fresh CA + installs
```

If both checks pass and you still see the error, the affected site has **HSTS preload + cert pinning** (banking sites, Google's own products like accounts.google.com under specific conditions, etc.) and will reject any MITM cert. Workaround: add the host to `passthrough_hosts` in config.json — browser hits it directly with the real cert, mhrv-rs doesn't intercept.

Drop the actual host from your log if you want me to confirm which category it falls in.

---
<sub>[reply via Anthropic Claude | reviewed by @therealaleph]</sub>

**Comment by presolve0** [comment]
@therealaleph How to fix this issue on windows?

**Comment by therealaleph** [comment]
Reviewed via Anthropic Claude.

@presolve0 — On Windows, mhrv-rs auto-installs the MITM CA into Windows's user trust store (`Current User\Trusted Root Certification Authorities`) on first run. If you see `Peer's certificate has an invalid signature`, the auto-install probably failed silently.

Manual install steps:

1. Find the CA cert: `%APPDATA%\mhrv-rs\ca.crt` (or wherever your config dir is — `mhrv-rs --print-config-dir`).
2. Double-click `ca.crt` → **Install Certificate** → **Current User** → **Place all certificates in the following store** → **Trusted Root Certification Authorities** → Next → Finish → confirm the security warning.
3. Restart your browser (Chrome/Edge picks up the user trust store automatically; Firefox uses its own store and needs separate config).

For Firefox specifically: `about:preferences#privacy` → scroll to **Certificates** → **View Certificates** → **Authorities** → **Import** → select `ca.crt` → check "Trust this CA to identify websites".

If you're getting the error in a specific app (not browser), that app may pin its own certs and won't accept any user CA — that's expected behavior, those apps need to be added to `passthrough_hosts` in `config.json` so mhrv-rs doesn't MITM them.

**Comment by therealaleph** [comment]
Closing this for now because there has not been a follow-up from the original reporter for more than 7 days.

If this is still happening on the latest release, please comment here or reopen with the current version, mode, relevant config shape, and fresh logs.

---
Answered via LLM, Supervised @therealaleph
